theseus/optimizer/nonlinear/trust_region.py

```diff
@@ -0,0 +1,167 @@
+        step_size: float = 1.0,
+        **kwargs,
+    ):
+        super().__init__(
+            objective,
+            linear_solver_cls=linear_solver_cls,
+            vectorize=vectorize,
+            linearization_cls=linearization_cls,
+            linearization_kwargs=linearization_kwargs,
+            linear_solver_kwargs=linear_solver_kwargs,
+            abs_err_tolerance=abs_err_tolerance,
+            rel_err_tolerance=rel_err_tolerance,
+            max_iterations=max_iterations,
+            step_size=step_size,
+            **kwargs,
+        )
+        self._trust_region: torch.Tensor = None
+        self._at_trust_boundary_idx: torch.Tensor = None
+
+    def reset(
+        self,
+        trust_region_init: float = 0.5,
+        **kwargs,
+    ) -> None:
+        super().reset(**kwargs)
+        self._trust_region = trust_region_init * torch.ones(
+            self.objective.batch_size,
+            1,
+            device=self.objective.device,
+            dtype=self.objective.dtype,
+        )
+        self._at_trust_boundary_idx = None
+
+    # Return the computed delta and, optionally, the indices that with the
+    # steps that are exactly at the boundary of the trust region
+    @abc.abstractmethod
+    def _compute_delta_impl(self) -> Tuple[torch.Tensor, Optional[torch.Tensor]]:
+        pass
+
+    def compute_delta(self, **kwargs) -> torch.Tensor:
+        # Storing the indices at the trust boundary so that we can update the
+        # trust region inside self.complete_step()
+        delta, self._trusted_step_idx = self._compute_delta_impl()
+        if self._trusted_step_idx is None:
+            self._trusted_step_idx = (
+                TrustRegion._squared_norm(delta) <= self._trust_region**2
+            )
+
+        # Storing the indices of delta exactly at the trust boundary.
+        # If the step is at the boundary and the predicted error reduction agrees
+        # with real reduction, we can increase the trust region
+        # in self.complete_step()
+        delta, self._at_trust_boundary_idx = self._compute_delta_impl()
+        if self._at_trust_boundary_idx is None:
+            self._at_trust_boundary_idx = (
+                TrustRegion._squared_norm(delta) <= self._trust_region**2
+            )
+        return delta
+
+    @staticmethod
+    def _squared_norm(tensor: torch.Tensor, keepdim: bool = True) -> torch.Tensor:
+        return (tensor**2).sum(dim=1, keepdim=keepdim)
+
+    # Computes m_k(delta) as defined in Eq. (4.2) of the above reference (p. 68)
+    def _predicted_error(
+        self, previous_error: torch.Tensor, delta: torch.Tensor
+    ) -> torch.Tensor:
+        linearization = self.linear_solver.linearization
+
+        # Note that B = |A^t@ A| so that p^tBp := delta^t A^t A delta = ||Adelta||^2
+        Adelta = linearization.Av(delta)
+        grad = -linearization.Atb.squeeze(2)
+        delta_dot_grad = (delta * grad).sum(dim=1)
+
+        return (
```

Comment:
Looks like the predicted error does not apply to LM method

Reply:
Yeah, I noticed it's a bit different in the current code. Will need to look into how to unify these later.